### PR TITLE
Allow executor tracing to flow across distinct executors

### PR DESCRIPTION
**Pull request by tylerbenson**
Otherwise any async work is lost after the first executor jump.

Also add some netty executors to the whitelist.

### CODE REVIEW

**Comment by tylerbenson** [comment]
@htmldoug, thanks for your patience.  This should resolve the issue you're seeing with #432.  You can test with the artifact saved in the build job on CircleCI, or you can wait till we merge to master and use the latest snapshot.

Let me know how it goes.

**Comment by htmldoug** [comment]
Thanks, @karkum-rally and I will give it a try as soon as we get a sec.

**Comment by karkum-rally** [comment]
@tylerbenson thanks for the quick fixes!

I tried the snapshot from https://8103-89221572-gh.circle-artifacts.com/0/home/circleci/dd-trace-java/libs/dd-java-agent-0.13.0-SNAPSHOT.jar (f49f3d70). 

First I tried @htmldoug's sample app and saw what I was expecting (shared `trace_id` between both spans and the correct `childOf` relationship):

```
[
  {
    "type": "web",
    "error": 0,
    "meta": {
      "http.status_code": "200",
      "component": "netty",
      "span.kind": "server",
      "http.url": "http://localhost:9000/post",
      "peer.hostname": "localhost",
      "env": "dev",
      "peer.port": "50470",
      "thread.name": "netty-event-loop-3",
      "http.method": "POST",
      "thread.id": "24",
      "span.type": "web"
    },
    "metrics": {
      "_sample_rate": 1,
      "_sampling_priority_v1": 1
    },
    "duration": 559823334,
    "resource": "POST /post",
    "service": "datadog_test_app",
    "name": "netty.request",
    "parent_id": 0,
    "start": 1534461267502882000,
    "trace_id": 7451078316826159000,
    "span_id": 2907457602815171000
  },
  {
    "type": "web",
    "error": 0,
    "meta": {
      "http.status_code": "200",
      "component": "play-action",
      "span.kind": "server",
      "http.url": "/post",
      "env": "dev",
      "thread.name": "application-akka.actor.default-dispatcher-6",
      "http.method": "POST",
      "thread.id": "42",
      "span.type": "web"
    },
    "metrics": {},
    "duration": 21079311,
    "resource": "POST /post",
    "service": "datadog_test_app",
    "name": "play.request",
    "parent_id": 2907457602815171000,
    "start": 1534461267954179300,
    "trace_id": 7451078316826159000,
    "span_id": 3409457607577213400
  }
]
```

I also integrated this snapshot with one of our applications and saw traces across several hops! 
screen shot 2018-08-16 at 4 39 08 pm

Here, the request path is: combinator -> missions -> user-settings, which is correctly shown above (with a mongo read in between).

However, I see some weird behavior in the Services page UI with this change:
screen shot 2018-08-16 at 4 45 17 pm

The resource list at the bottom of the page includes resources that aren't actually being served by combinator. These are instead endpoints that combinator itself is calling out to. For example, `/rest/users/?/userTags` is an endpoint in another service entirely that combinator calls. I'm not sure if this issue is related to this change, but I hadn't seen this behavior before.

**Comment by tylerbenson** [comment]
@karkum-rally Looking at those specific traces, it appears the root span is coming from `AsyncHttpClient`, not a regular web application root.  I assume that this async http request is not being linked up properly with the original incoming request.

Can you share how details on the execution flow of this application from when netty receives the request till making the async client request?  (Another sample app that repros the problem would also be awesome.  Made troubleshooting the earlier issue so much easier.)

**Comment by karkum-rally** [comment]
@tylerbenson here you go: https://github.com/htmldoug/datadog-netty4-failing/pull/1

The description there has the issue we're seeing. The steps to reproduce are the same as the ones https://github.com/DataDog/dd-trace-java/issues/432.

**Comment by tylerbenson** [comment]
Thanks. I’ll take a look next week.

**Comment by tylerbenson** [comment]
@karkum-rally @htmldoug Can you try this build: https://8375-89221572-gh.circle-artifacts.com/0/home/circleci/dd-trace-java/libs/dd-java-agent-0.13.0-SNAPSHOT.jar?

**Comment by karkum-rally** [comment]
@tylerbenson thanks, I'll look into this today

**Comment by karkum-rally** [comment]
@tylerbenson I first verified the sample app and confirmed that it was able to link the requests correctly. I also integrated that snapshot with one of our apps and the traces look much more accurate!

screen shot 2018-08-21 at 4 52 11 pm

The only question I have now is around the resource list on the Service home page (/apm/services/<serviceName>):
screen shot 2018-08-21 at 5 00 58 pm

These resources are not served by this service, instead these are resources that this service calls out to. I'm not sure if my expectation is incorrect, but I thought this list would only show the former. Can you please confirm?

I will work with @htmldoug tomorrow to see if he wants to run any tests on one of his apps, but I think this looks good so far!

**Comment by tylerbenson** [comment]
@karkum-rally @htmldoug This might be a good point to move this discussion to a support ticket.  That way you can be more specific.  Lets call the original problems solved and consider this a separate issue.

**Comment by karkum-rally** [comment]
@tylerbenson sounds good, thanks! Sorry for piling on in my last comment, I will follow up in a support ticket.

**Comment by stuffandthings** [comment]
Good morning,

Just checking in on the status of this as I believe we're encountering this with our tracing instrumentation on our Scala (play framework) application.

I'd be happy to contribute/help with this PR if I can! Please let me know.

**Comment by htmldoug** [comment]
I'm also waiting for this to be merged before turning on APM in prod to finish our evaluation. Snapshots in prod makes me uneasy.

File: dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/ThreadPoolExecutorInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+@Slf4j
+@AutoService(Instrumenter.class)
+public class ThreadPoolExecutorInstrumentation extends Instrumenter.Default {
+
+  public ThreadPoolExecutorInstrumentation() {
+    super(ExecutorInstrumentation.EXEC_NAME);
```
Comment (realark) [inline dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/ThreadPoolExecutorInstrumentation.java]: I think we need to refresh this branch with master.

**Pull request merged**